cpp/DataStorm/sampleFilter/Reader.cpp

```diff
@@ -4,40 +4,57 @@
 #include <Ice/Ice.h>
 
 #include <iostream>
-#include <optional>
 #include <string>
 
+#include "TemperatureRange.h"
+
 using namespace std;
 
-int
-main(int argc, char* argv[])
+void
+printSample(const DataStorm::Sample<string, float>& sample)
 {
-    try
-    {
-        // Instantiates DataStorm node.
-        DataStorm::Node node{argc, argv};
-
-        // Instantiates the "hello" topic. The topic uses strings for keys and values.
-        DataStorm::Topic<string, string> topic{node, "hello"};
-
-        // Configure readers to never clear the history. We want to receive all the samples written by the writers.
-        topic.setReaderDefaultConfig({std::nullopt, std::nullopt, DataStorm::ClearHistoryPolicy::Never});
-
-        // Instantiate the reader for the key "foo". The reader uses the predefined _regex sample filter and the
-        // "good.*" regular expression as the criteria.
-        auto reader = DataStorm::makeSingleKeyReader(topic, "foo", DataStorm::Filter<string>("_regex", "good.*"));
-
-        // Get the 2 samples starting with good published by the writer.
-        auto sample = reader.getNextUnread();
-        cout << sample.getKey() << " says " << sample.getValue() << "!" << endl;
+    cout << "sample: key=" << sample.getKey() << ", value=" << sample.getValue() << std::endl;
+}
 
-        sample = reader.getNextUnread();
-        cout << sample.getKey() << " says " << sample.getValue() << "!" << endl;
-    }
-    catch (const std::exception& ex)
+void
+printSamples(const vector<DataStorm::Sample<string, float>>& samples)
+{
+    for (const auto& sample : samples)
     {
-        cerr << ex.what() << endl;
-        return 1;
+        printSample(sample);
     }
+}
+
+int
+main(int argc, char* argv[])
+{
+    // CtrlCHandler must be created before the communicator is created or any other threads are started.
+    Ice::CtrlCHandler ctrlCHandler;
+
+    // Instantiates DataStorm node.
+    DataStorm::Node node{argc, argv};
+
+    ctrlCHandler.setCallback(
+        [&ctrlCHandler, &node](int)
+        {
+            std::cout << "Shutting down..." << std::endl;
+            node.shutdown();
+            // Reset the callback to nullptr to avoid calling it again.
+            ctrlCHandler.setCallback(nullptr);
```

Comment:
Is this necessary? I haven't seen it in other demos, and `setCallback` locks a non-reentrant mutex internally.
Locking mutexes in a signal handler always feels a little dicey to me.

Reply:
Copied from one of the demos, but seems we don't have this in most of the demos.

Reply:
Removed